week-1-ts-ecosystem/w01e03-object-oriented-structure/src/model/cart.ts

```diff
@@ -0,0 +1,71 @@
+import { v4 } from 'uuid';
+import { CartType, Currency, ICartItem } from './cart.model';
+
+export class CartItem implements ICartItem {
+    id: string
+    constructor(
+        public name: string,
+        public count: number,
+        public price: number,
+        public currency: Currency,
+        public type: CartType) {
+            this.id = v4();
+        }
+
+        update(count: number): void {
+            if(count > 0) {
+                this.count = count;
+            }
+        }
+}
+
+export class Cart {
+    private items: Map<string, CartItem> = new Map();
+    constructor(public type: CartType) {}
+
+    add(item: CartItem): string {
+        if(!(item instanceof CartItem)) {
+            this.throwCartItemTypeError();
```

Comment:
a jakbyś przekazywał `item` do środka metody `throwCartItemTypeError` ?
Pochowałbyś @Dam1anM wtedy te `if` y

bo to tam w środku byś to sprawdzał.

Wtedy kod wyglądałby tak:
```typescript
this.throwIfNotCartItemType(item);
this.throwIfNotCartTypeError(item);
this.items.set(item.id, item);
return item.id;
```

z perspektywy miejsca rzucania wyjątku nie ma to znaczenia że to w środku tych metod będziesz sprawdzał warunki.

Reply:
Tutaj mam dylemat. 
Z jednej strony ten kod tutaj (Twój) wygląda schludniej i do mnie to bardziej przemawia, natomiast kiedyś oglądałem wykład Robert C. Martin i on tam powiedział takie zdanie, które mi mocno utkwiło w głowie, a tyczyło się przesyłania do metody wartości będącą flagą. Tutaj to nie jest dosłownie flaga ale **item** jaki miałbym przesłać do metody spełniałby taką rolę w tej sytuacji. Robert C. Martin jest zdania, że to nie dobre podejście, bo przez to ten kod jest troche zaciemniony, a dodatkowo metody stają się bardziej skomplikowane, bo do tej pory ich zadanie było jedno -> rzuć błędem, a teraz mamy już dwa "branche" do sprawdzenia tj. błąd jest rzucany albo nie jest. Noi druga kwestia, czyli debagując taki kod musimy wejść do wszystkich metod i wewnątrz nich sprawdzać czy warunek został spełniony, a gdy mamy zapis taki jak u mnie to wchodzimy tylko tam gdzie warunek się spełnił.

Ciekaw jestem Twoje opinii @michaljabi

Reply:
Ja tutaj bazuje na nazwie.

Zobacz jest wyraźnie napisane `throwIfNotCartItemType` i przekazujesz coś.
Więc jakby sugerujesz że coś rzuci jeśli coś nie będzie spełnione.

Wiem że ta nazwa "może kłamać".

Ale w mojej opinii rzucanie błędów w kodzie polega na tym że możesz je rzucić z dowolnie głębokiego miejsca - właśnie po to żeby złapać to gdzieś wyżej @Dam1anM .

I fakt - masz rację z tym że to jest "zasłonięte" - ale może się okazać dzięki temu bardziej "re-używalne).
W innym przypadku - pewnie to nie byłoby dobry pomysł. Jednak tutaj misja jest taka żeby rzucić coś na niespełnienie warunku. Co do debugowania - to nie ma to dla mnie większego znaczenia że jak odpalę debuger to mam jedno więcej miejsce do przeskoczenia w razie czego. 

Dodatkowo widzę tutaj chyba nadmiarowe sprawdzenie.

jak zapiszemy:
```typescript
add(item: CartItem) {}
```

to sam `typescript` sprawdza przecież poprawność przekazanego `item` więc nie musisz sprawdzać `insanceof` ?
to mamy odchaczone.